tests/view/observer/selectionobserver.js

```diff
@@ -124,6 +130,71 @@ describe( 'SelectionObserver', () => {
 		changeDomSelection();
 	} );
 
+	it( 'should warn and not enter infinite loop', ( done ) => {
+		let counter = 30;
+
+		const viewFoo = viewDocument.getRoot().getChild( 0 ).getChild( 0 );
+		viewDocument.selection.addRange( ViewRange.createFromParentsAndOffsets( viewFoo, 0, viewFoo, 0 ) );
+
+		listenter.listenTo( viewDocument, 'selectionChange', () => {
+			counter--;
+
+			if ( counter > 0 ) {
+				setTimeout( changeDomSelection );
+			} else {
+				throw( 'Infinite loop!' );
+			}
+		} );
+
+		testUtils.sinon.stub( log, 'warn', ( msg ) => {
+			expect( msg ).to.match( /^selectionchange-inifite-loop/ );
+			done();
+
+			// Prevent "done() called multiple times" error.
+			done = () => {};
+		} );
+
+		changeDomSelection();
+	} );
+
+	it( 'should not be thread as an infinite loop if the position is different', ( done ) => {
+		let counter = 30;
+
+		const viewFoo = viewDocument.getRoot().getChild( 0 ).getChild( 0 );
+		viewDocument.selection.addRange( ViewRange.createFromParentsAndOffsets( viewFoo, 0, viewFoo, 0 ) );
+
+		listenter.listenTo( viewDocument, 'selectionChange', () => {
+			counter--;
+
+			if ( counter > 0 ) {
+				setTimeout( () => changeCollapsedDomSelection( counter ) );
+			} else {
+				done();
+			}
+		} );
+
+		changeCollapsedDomSelection( counter );
+	} );
+
+	it( 'should not be thread as an infinite loop if it is less then 3 times', ( done ) => {
```

Comment:
Which reminds me that we haven't clarified this in https://github.com/ckeditor/ckeditor5-design/wiki/Code-Style. Let's pretend we haven't seen this.